dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/RepeatableStageParamsHelper.java

```diff
@@ -0,0 +1,264 @@
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.analytics.util;
+
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import org.hisp.dhis.common.RepeatableStageParams;
+import org.hisp.dhis.common.exception.InvalidRepeatableStageParamsException;
+import org.hisp.dhis.period.Period;
+import org.hisp.dhis.period.RelativePeriodEnum;
+import org.hisp.dhis.period.RelativePeriods;
+import org.hisp.dhis.util.DateUtils;
+
+public class RepeatableStageParamsHelper
+{
+    private static final String PS_INDEX_REGEX = "\\[-?\\d+\\]";
+
+    private static final String PS_ASTERISK_REGEX = "\\[\\*\\]";
+
+    private static final String PS_INDEX_COUNT_REGEX = "\\[-?\\d+,\\s*\\d+\\]";
+
+    private static final String PS_INDEX_COUNT_START_DATE_END_DATE_REGEX = "\\[-?\\d+,\\s*\\d+,\\s*\\d{4}\\-(0[1-9]|1[012])\\-(0[1-9]|[12][0-9]|3[01]),\\s*\\d{4}\\-(0[1-9]|1[012])\\-(0[1-9]|[12][0-9]|3[01])\\]";
+
+    private static final String PS_INDEX_COUNT_RELATIVE_PERIOD_REGEX = "\\[-?\\d+,\\s*\\d+,\\s*\\w+\\]";
+
+    private static final String PS_START_DATE_END_DATE_REGEX = "\\[\\d{4}\\-(0[1-9]|1[012])\\-(0[1-9]|[12][0-9]|3[01]),\\s*\\d{4}\\-(0[1-9]|1[012])\\-(0[1-9]|[12][0-9]|3[01])\\]";
+
+    private static final String PS_RELATIVE_PERIOD_REGEX = "\\[\\w+\\]";
+
+    private static final Pattern[] PS_PARAMS_PATTERN_LIST = {
+        Pattern.compile( PS_INDEX_REGEX ),
+        Pattern.compile( PS_ASTERISK_REGEX ),
+        Pattern.compile( PS_INDEX_COUNT_REGEX ),
+        Pattern.compile( PS_INDEX_COUNT_START_DATE_END_DATE_REGEX ),
+        Pattern.compile( PS_INDEX_COUNT_RELATIVE_PERIOD_REGEX ),
+        Pattern.compile( PS_START_DATE_END_DATE_REGEX ),
+        Pattern.compile( PS_RELATIVE_PERIOD_REGEX )
+    };
+
+    private RepeatableStageParamsHelper()
+    {
+    }
+
+    public static RepeatableStageParams getRepeatableStageParams( String dimension )
+        throws InvalidRepeatableStageParamsException
+    {
+        Pattern pattern = Arrays.stream( PS_PARAMS_PATTERN_LIST )
+            .filter( p -> p.matcher( dimension ).find() )
+            .findFirst()
+            .orElse( Pattern.compile( "" ) );
+
+        final Matcher matcher = pattern.matcher( dimension );
+
+        List<String> tokens;
+
+        switch ( pattern.toString() )
+        {
+        case PS_ASTERISK_REGEX:
+
+            return getRepeatableStageParams( 0, Integer.MAX_VALUE );
+        case PS_INDEX_COUNT_REGEX:
+            tokens = getMatchedRepeatableStageParamTokens( matcher, 2 );
+
+            return getRepeatableStageParams( Integer.parseInt( tokens.get( 0 ) ), Integer.parseInt( tokens.get( 1 ) ) );
```

Comment:
We are assuming that `tokens.get( 0 )` will always be a number. Just checking if this will be always the case.
The same applies to the code below.